utility.lua

```diff
@@ -70,38 +70,48 @@ local get_placer = function(player_name)
 end
 
 -- Used to remove items from the target block and put it into the hopper's inventory
-hopper.take_item_from = function(hopper_pos, target_pos, target_node, target_inventory_name)
-	if target_inventory_name == nil then
-		return
-	end
+hopper.take_item_from = function(hopper_pos, target_pos, target_node, target_inv_info)
 	local target_def = minetest.registered_nodes[target_node.name]
 	if not target_def then
 		return
 	end
 
+	local target_inv_name = target_inv_info["inventory_name"]
+	local target_get_inv  = target_inv_info["get_inventory"]
+
 	--hopper inventory
-	local hopper_meta = minetest.get_meta(hopper_pos);
+	local hopper_meta = minetest.get_meta(hopper_pos)
 	local hopper_inv = hopper_meta:get_inventory()
 	local placer = get_placer(hopper_meta:get_string("placer"))
 
 	--source inventory
-	local target_inv = minetest.get_meta(target_pos):get_inventory()
-	local target_inv_size = target_inv:get_size(target_inventory_name)
-	if target_inv:is_empty(target_inventory_name) == false then
+	local target_inv
+	if target_get_inv then
+		target_inv = target_get_inv(target_pos)
+		if not target_inv then
+			minetest.log("error","No inventory from api get_inventory function: " ..
+				target_node.name .. " on " .. vector.to_string(target_pos))
+			return false
+		end
+	else
+		target_inv = minetest.get_meta(target_pos):get_inventory()
+	end
+	local target_inv_size = target_inv:get_size(target_inv_name)
```

Comment:
This code is duplicated (`hopper.send_item_to`). Suggestion: move to a separate helper function:

```lua
--- @returns inventory (InvRef), listname (string)
---          ^ may be `nil` when `get_inventory()` returns an unexpected value
local function hoper_get_inventory(node_pos, inv_info)
	local list_name = inv_info.inventory_name
	local get_inventory_fn = inv_info.get_inventory

	local inventory
	if get_inventory_fn then
		inventory = get_inventory_fn(target_pos)
		if not inventory then
			minetest.log("error","No inventory from api get_inventory function: " ..
				target_node.name .. " on " .. vector.to_string(node_pos))
		end
	else
		-- Fallback
		inventory = minetest.get_meta(node_pos):get_inventory()
	end
	return inventory, list_name
end


-- To use (using existing variable names)
local target_inv, target_inv_name = hoper_get_inventory(target_pos, target_inv_info)
if not target_inv then
	return false
end
local target_inv_size = target_inv:get_size(target_inv_name)
```

Reply:
Applied.